File: CON BDD/EUREKABANK_RESTFULL_JAVA_GR06/08. SERVIDOR/eurekabank_restful_java_gr06/src/main/java/ec/edu/monster/persistencia/ClienteDAO.java
package ec.edu.monster.persistencia;

import ec.edu.monster.modelo.ClienteResumen;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

/** Acceso a datos de la tabla `cliente`. */
public class ClienteDAO {

    /** Lista todos los clientes registrados (solo lectura). */
    public List<ClienteResumen> listarTodos() throws SQLException {
        String sql = "SELECT chr_cliecodigo, chr_cliedni, "
                   + "vch_clienombre, vch_cliepaterno, vch_cliematerno "
                   + "FROM cliente ORDER BY vch_cliepaterno, vch_cliematerno, vch_clienombre";
        List<ClienteResumen> lista = new ArrayList<>();
        Connection cn = null;
        PreparedStatement ps = null;
        ResultSet rs = null;
        try {
            cn = ConexionBD.conectar();
            ps = cn.prepareStatement(sql);
            rs = ps.executeQuery();
            while (rs.next()) {
                ClienteResumen c = new ClienteResumen();
                c.setCodigo(rs.getString("chr_cliecodigo"));
                c.setDni(rs.getString("chr_cliedni"));
                c.setNombre((rs.getString("vch_clienombre") + " "
                        + rs.getString("vch_cliepaterno") + " "
                        + rs.getString("vch_cliematerno")).trim());
                lista.add(c);
            }
            return lista;
        } finally {
            ConexionBD.desconectar(rs);
            ConexionBD.desconectar(ps);
            ConexionBD.desconectar(cn);
        }
    }

    /**
     * Inserta un cliente con código autogenerado (MAX+1, 5 dígitos).
     * @return el código del cliente creado.
     */
    public String insertar(String paterno, String materno, String nombre,
                           String dni, String ciudad, String direccion,
                           String telefono, String email) throws SQLException {
        Connection cn = null;
        PreparedStatement ps = null;
        ResultSet rs = null;
        try {
            cn = ConexionBD.conectar();
            String codigo;
            try (PreparedStatement pm = cn.prepareStatement(
                    "SELECT LPAD(COALESCE(MAX(CAST(chr_cliecodigo AS UNSIGNED)),0)+1,5,'0') "
                    + "FROM cliente");
                 ResultSet rm = pm.executeQuery()) {
                rm.next();
                codigo = rm.getString(1);
            }
            String sql = "INSERT INTO cliente (chr_cliecodigo, vch_cliepaterno, "
                       + "vch_cliematerno, vch_clienombre, chr_cliedni, vch_clieciudad, "
                       + "vch_cliedireccion, vch_clietelefono, vch_clieemail) "
                       + "VALUES (?,?,?,?,?,?,?,?,?)";
            ps = cn.prepareStatement(sql);
            ps.setString(1, codigo);
            ps.setString(2, paterno);
            ps.setString(3, materno);
            ps.setString(4, nombre);
            ps.setString(5, dni);
            ps.setString(6, ciudad);
            ps.setString(7, direccion);
            ps.setString(8, telefono);
            ps.setString(9, email);
            ps.executeUpdate();
            return codigo;
        } finally {
            ConexionBD.desconectar(rs);
            ConexionBD.desconectar(ps);
            ConexionBD.desconectar(cn);
        }
    }

    /** Verifica si existe un cliente por su código. */
    public boolean existe(String codigo) throws SQLException {
        Connection cn = null;
        PreparedStatement ps = null;
        ResultSet rs = null;
        try {
            cn = ConexionBD.conectar();
            ps = cn.prepareStatement(
                    "SELECT 1 FROM cliente WHERE chr_cliecodigo = ?");
            ps.setString(1, codigo);
            rs = ps.executeQuery();
            return rs.next();
        } finally {
            ConexionBD.desconectar(rs);
            ConexionBD.desconectar(ps);
            ConexionBD.desconectar(cn);
        }
    }
}

File: CON BDD/EUREKABANK_REST_DOTNET_GR06/08. SERVIDOR/Data/ClienteDAO.cs
using Microsoft.Data.SqlClient;
using SERVIDOR.Models;

namespace SERVIDOR.Data;

public class ClienteDAO
{
    public List<ClienteResumen> ListarTodos()
    {
        const string sql = """
            SELECT chr_cliecodigo, chr_cliedni,
                   vch_clienombre, vch_cliepaterno, vch_cliematerno
            FROM cliente ORDER BY vch_cliepaterno, vch_cliematerno, vch_clienombre
            """;

        var lista = new List<ClienteResumen>();
        using var cn = ConexionBD.CrearConexion();
        using var ps = new SqlCommand(sql, cn);
        cn.Open();
        using var rs = ps.ExecuteReader();
        while (rs.Read())
        {
            lista.Add(new ClienteResumen
            {
                Codigo = rs.GetString(0),
                Dni = rs.GetString(1),
                Nombre = $"{rs.GetString(2)} {rs.GetString(3)} {rs.GetString(4)}".Trim()
            });
        }
        return lista;
    }

    public string Insertar(string paterno, string materno, string nombre,
                           string dni, string ciudad, string direccion,
                           string telefono, string email)
    {
        using var cn = ConexionBD.CrearConexion();
        cn.Open();

        string codigo;
        using (var pm = new SqlCommand(
            "SELECT RIGHT('00000' + CAST(COALESCE(MAX(CAST(chr_cliecodigo AS INT)), 0) + 1 AS VARCHAR(5)), 5) FROM cliente", cn))
        {
            codigo = pm.ExecuteScalar()!.ToString()!;
        }

        const string sql = """
            INSERT INTO cliente (chr_cliecodigo, vch_cliepaterno,
                   vch_cliematerno, vch_clienombre, chr_cliedni, vch_clieciudad,
                   vch_cliedireccion, vch_clietelefono, vch_clieemail)
            VALUES (@codigo, @paterno, @materno, @nombre, @dni, @ciudad, @direccion, @telefono, @email)
            """;
        using var ps = new SqlCommand(sql, cn);
        ps.Parameters.AddWithValue("@codigo", codigo);
        ps.Parameters.AddWithValue("@paterno", paterno);
        ps.Parameters.AddWithValue("@materno", materno);
        ps.Parameters.AddWithValue("@nombre", nombre);
        ps.Parameters.AddWithValue("@dni", dni);
        ps.Parameters.AddWithValue("@ciudad", ciudad);
        ps.Parameters.AddWithValue("@direccion", direccion);
        ps.Parameters.AddWithValue("@telefono", telefono);
        ps.Parameters.AddWithValue("@email", email);
        ps.ExecuteNonQuery();
        return codigo;
    }

    public bool Existe(string codigo)
    {
        using var cn = ConexionBD.CrearConexion();
        using var ps = new SqlCommand(
            "SELECT 1 FROM cliente WHERE chr_cliecodigo = @codigo", cn);
        ps.Parameters.AddWithValue("@codigo", codigo);
        cn.Open();
        return ps.ExecuteScalar() != null;
    }
}
